pallets/hyperdrive/src/tezos.rs

```diff
@@ -0,0 +1,570 @@
+use std::marker::PhantomData;
+
+use derive_more::{Display, From};
+use frame_support::once_cell::race::OnceBox;
+use frame_support::Parameter;
+use sp_core::bounded::BoundedVec;
+use sp_core::ConstU32;
+use sp_runtime::traits::Member;
+use sp_std::prelude::*;
+use sp_std::str::FromStr;
+use tezos_core::types::encoded::Address as TezosAddress;
+use tezos_core::Error as TezosCoreError;
+use tezos_michelson::micheline::primitive_application::PrimitiveApplication;
+use tezos_michelson::micheline::Micheline;
+use tezos_michelson::michelson::data;
+use tezos_michelson::michelson::data::{try_int, try_string, Bytes, Data, Int, Pair, Sequence};
+use tezos_michelson::michelson::types::{
+    address, bool as bool_type, bytes, nat, option, pair, set, string,
+};
+use tezos_michelson::Error as TezosMichelineError;
+
+use pallet_acurast::{JobIdSequence, JobRegistration, MultiOrigin, Schedule};
+use pallet_acurast_marketplace::{JobRequirements, PlannedExecution, RegistrationExtra};
+
+use crate::types::{MessageParser, RawAction};
+use crate::Error;
+use crate::{Config, ParsedAction};
+
+pub struct TezosParser<Reward, Balance, ParsableAccountId, AccountId, Extra>(
+    PhantomData<(Reward, Balance, ParsableAccountId, AccountId, Extra)>,
+);
+
+impl<Reward, Balance, ParsableAccountId, AccountId, Extra> MessageParser<AccountId, Extra>
+    for TezosParser<Reward, Balance, ParsableAccountId, AccountId, Extra>
+where
+    ParsableAccountId: FromStr + Into<AccountId>,
+    Extra: From<RegistrationExtra<Reward, Balance, AccountId>>,
+    Reward: Parameter + Member + TryFrom<Vec<u8>>,
+    Balance: From<u128>,
+{
+    type Error = ValidationError;
+    fn parse(encoded: &[u8]) -> Result<ParsedAction<AccountId, Extra>, ValidationError> {
+        let (action, origin, payload) = parse_message(encoded)?;
+
+        Ok(match action {
+            RawAction::RegisterJob => {
+                let payload: Vec<u8> = (&payload).into();
+                let (job_id_sequence, registration) = parse_job_registration_payload::<
+                    Reward,
+                    Balance,
+                    ParsableAccountId,
+                    AccountId,
+                    Extra,
+                >(payload.as_slice())?;
+
+                ParsedAction::RegisterJob(
+                    (to_multi_origin(&origin)?, job_id_sequence),
+                    registration,
+                )
+            }
+        })
+    }
+}
+
+impl<T: Config, I> From<ValidationError> for Error<T, I> {
+    fn from(_: ValidationError) -> Self {
```

Comment:
The `ValidationError` is lost here. We should at least log it.

Reply:
we now emit events for error cases (and not fail extrinsic `submit_message`)